src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java

```diff
@@ -249,27 +260,56 @@ private CompletableFuture<Void> registerFuture() {
   }
 
   /**
-   * Kick off a flush job and distribute the tasks if one of the following conditions is met:
-   * <li>Flush is forced by the users
-   * <li>One or more buffers have reached the flush size
-   * <li>Periodical background flush when a time interval has reached
+   * Kick off a flush job and distribute the tasks. The flush service behaves differently based on
+   * the max chunks in blob:
+   *
+   * <ul>
+   *   <li>The max chunks in blob is not 1 (interleaving is allowed), every channel will be flushed
+   *       together if one of the following conditions is met:
+   *       <ul>
+   *         <li>Flush is forced by the users
+   *         <li>One or more buffers have reached the flush size
+   *         <li>Periodical background flush when a time interval has reached
+   *       </ul>
+   *   <li>The max chunks in blob is 1 (interleaving is not allowed), a channel will be flushed if
+   *       one of the following conditions is met:
+   *       <ul>
+   *         <li>Flush is forced by the users
+   *         <li>One or more buffers with the same target table as the channel have reached the
+   *             flush size
+   *         <li>Periodical background flush of the target table when a time interval has reached
+   *       </ul>
+   * </ul>
    *
    * @param isForce
    * @return Completable future that will return when the blobs are registered successfully, or null
    *     if none of the conditions is met above
    */
   CompletableFuture<Void> flush(boolean isForce) {
-    long timeDiffMillis = System.currentTimeMillis() - this.lastFlushTime;
-
-    if (isForce
-        || (!DISABLE_BACKGROUND_FLUSH
-            && !isTestMode()
-            && (this.isNeedFlush
-                || timeDiffMillis
-                    >= this.owningClient.getParameterProvider().getCachedMaxClientLagInMs()))) {
+    long flushStartTime = System.currentTimeMillis();
+
+    Set<String> tablesToFlush =
+        this.channelCache.keySet().stream()
+            .filter(
+                key ->
+                    isForce
+                        || (this.channelCache.getLastFlushTime(key).isPresent()
+                            && flushStartTime - this.channelCache.getLastFlushTime(key).get()
+                                >= this.owningClient
+                                    .getParameterProvider()
+                                    .getCachedMaxClientLagInMs())
+                        || this.channelCache.getNeedFlush(key))
+            .collect(Collectors.toSet());
+
+    // Flush every table together when it's interleaving chunk is allowed
+    if (this.owningClient.getParameterProvider().getMaxChunksInBlobAndRegistrationRequest() != 1
+        && !tablesToFlush.isEmpty()) {
```

Comment:
Can we do the check before populating `tablesToFlush`?

Reply:
Discussed offline. Preserve the client level `lastFlushTime` and `isNeedFlush` to avoid checking table level flush info when interleaving is enabled which might cause performance change. Preserve old logging format when interleaving is enabled to avoid logging too much information.

cc: @sfc-gh-hmadan